include/lo2s/perf/reader.hpp

```diff
@@ -0,0 +1,368 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <set>
+#include <variant>
+
+#include <lo2s/config.hpp>
+#include <lo2s/error.hpp>
+#include <lo2s/execution_scope.hpp>
+#include <lo2s/perf/tracepoint/format.hpp>
+#include <lo2s/perf/util.hpp>
+
+#ifndef USE_HW_BREAKPOINT_COMPAT
+extern "C"
+{
+#include <linux/hw_breakpoint.h>
+}
+#else
+extern "C"
+{
+#include <sys/types.h>
+#include <sys/wait.h>
+}
+#endif
+
+namespace lo2s
+{
+namespace perf
+{
+
+enum class Availability
+{
+    UNAVAILABLE,
+    SYSTEM_MODE,
+    PROCESS_MODE,
+    UNIVERSAL
+};
+
+class PerfEventInstance;
+
+/**
+ * Base class for all Event types
+ * contains common attributes
+ */
+class PerfEvent
+{
+public:
+    PerfEvent(bool enable_on_exec);
+    PerfEvent();
+
+    /**
+     * returns an opened instance of any PerfEvent object
+     * @param group_fd file descriptor of the leader event, omit if there is no leader
+     */
+    PerfEventInstance open(std::variant<Cpu, Thread> location, int group_fd = -1,
+                           int cgroup_fd = config().cgroup_fd);
+
+    perf_event_attr& get_attr()
+    {
+        return attr_;
+    }
+
+    double get_scale() const
+    {
+        return scale_;
+    }
+
+    void set_scale(double scale)
+    {
+        scale_ = scale;
+    }
+
+    std::string get_unit() const
+    {
+        return unit_;
+    }
+
+    void set_unit(std::string unit)
+    {
+        unit_ = unit;
+    }
+
+    friend bool operator==(const PerfEvent& lhs, const PerfEvent& rhs)
+    {
+        return !memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator<(const PerfEvent& lhs, const PerfEvent& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    ~PerfEvent()
+    {
+    }
+
+protected:
+    struct perf_event_attr attr_;
+    double scale_ = 1;
+    std::string unit_ = "#";
+};
+
+/**
+ * Contains predefined events, like the ones listed by --list-events
+ * @note define the type by either calling as_counter() or as_group_leader() after creating an
+ * object
+ */
+class PredefEvent : public PerfEvent
+{
+public:
+    PredefEvent();
+    PredefEvent(bool enable_on_exec, const std::string& ev_name);
+    PredefEvent(std::string name, perf_type_id type, std::uint64_t config,
+                std::uint64_t config1 = 0, std::set<Cpu> cpus = std::set<Cpu>());
```

Comment:
Check where cpus is coming from and ideally integrate that code directly into the Constructor.

Reply:
There are multiple ways to get cpus. Once through `Topology::instance().cpus()`, which takes cpus from `/sys/devices/system/cpu/online` and one which takes them from `/sys/bus/event_source/devices/cpus` or `-/cpumask`. 
It doesn't look like they can be combined.

Reply:
Use Topology::instance().cpus() when cpus or cpumask do not exist.

Use either cpus or cpumask when they do exist. An event cant have both a cpus and cpumask file.


Thie logic here is that cpus and cpumask specify that some event is not available on all CPUs and instead is only available on the cpus listed in that file (why there needs to be two different files, nobody knows). 

If neither file exist we expect the event to be available on every cpu, so use the list of all cpus from Topology.